### PR TITLE
Include dtype information

### LINKED ISSUE #25: Add tensor element type info
From @sbrugman:

Our concrete issue was with Pytorch (unexpectedly) converting tensors with only integers to float, which later in the program resulted in an error because it could not be used as an index. 
Another issue was changing size from 32 to 64 bit floats.

It's indeed the element type of the matrix.

There are multiple somewhat related issues:
https://discuss.pytorch.org/t/problems-with-target-arrays-of-int-int32-types-in-loss-functions/140/2
https://discuss.pytorch.org/t/why-pytorch-is-giving-me-hard-time-with-float-long-double-tensor/14678/6

The common denominator between dimensionality debugging is that both type and dimensionality are practically hidden from the user:

Screenshot 2021-09-20 at 20 00 33

```python
import numpy as np
import tsensor as ts

x = np.arange(6, dtype=np.float32)

with ts.explain(savefig="types.pdf"):
    print(x.dtype)
    print((x*x).dtype)
    print((np.sin(x)).dtype)
    print((x + np.arange(6)).dtype)
    print((np.multiply.outer(x, np.arange(2.0))).dtype)
    print((np.outer(x, np.arange(2.0))).dtype)
```

### PR DESCRIPTION
**Pull request by sbrugman**
Closes https://github.com/parrt/tensor-sensor/issues/25

### CODE REVIEW

**Comment by parrt** [comment]
Looking cool...I still need to think about how I'd viz this. Maybe the elem type goes in the box not outside, for example.

**Comment by sbrugman** [comment]
Discussion on that could go in the issue. With respect to implementation: this feature still needs testing,  mention in documentation/examples.

**Comment by sbrugman** [comment]
Failing tests seem [unrelated](https://stackoverflow.com/questions/56859803/modulenotfounderror-no-module-named-tools-nnwrap).

Two examples to for review.

| Example 1 |                                                     Legend                                                    |                                                                                                           No Legend |
|-----------|:-------------------------------------------------------------------------------------------------------------:|--------------------------------------------------------------------------------------------------------------------:|
| Dtypes    | [example-dtype-legend-1.pdf](https://github.com/parrt/tensor-sensor/files/7522256/example-dtype-legend-1.pdf) | [example-dtype-no-legend-1.pdf](https://github.com/parrt/tensor-sensor/files/7522258/example-dtype-no-legend-1.pdf) |
| Shape     | [example-shape-legend-1.pdf](https://github.com/parrt/tensor-sensor/files/7522260/example-shape-legend-1.pdf) | [example-shape-no-legend-1.pdf](https://github.com/parrt/tensor-sensor/files/7522262/example-shape-no-legend-1.pdf) |


| Example 2 |                                                     Legend                                                    |                                                                                                           No Legend |
|-----------|:-------------------------------------------------------------------------------------------------------------:|--------------------------------------------------------------------------------------------------------------------:|
| Dtypes    | [example-dtype-legend-2.pdf](https://github.com/parrt/tensor-sensor/files/7522257/example-dtype-legend-2.pdf) | [example-dtype-no-legend-2.pdf](https://github.com/parrt/tensor-sensor/files/7522259/example-dtype-no-legend-2.pdf) |
| Shape     | [example-shape-legend-2.pdf](https://github.com/parrt/tensor-sensor/files/7522261/example-shape-legend-2.pdf) | [example-shape-no-legend-2.pdf](https://github.com/parrt/tensor-sensor/files/7522263/example-shape-no-legend-2.pdf) |

**Comment by sbrugman** [comment]
Ready for review, legend improved over the last time we spoke

**Comment by parrt** [comment]
sounds good @sbrugman ...  I am writing my final exam and hoping to catch up shortly...

**Comment by parrt** [comment]
Trying it out now :)

File: tsensor/viz.py [hunk 1]
```diff
@@ -71,6 +91,62 @@ def __init__(self, statement, fontname, fontsize, dimfontname, dimfontsize,
...
+
+    @staticmethod
+    def _split_dtype_precision(s):
+        """Split the final integer part from a string"""
+        head = s.rstrip('0123456789')
+        tail = s[len(head):]
+        return head, tail
+
+    def _get_dtype_color_indices(self, dtype_name, dtype_precision):
+        if dtype_name not in self._dtype_name2color:
```
Comment (parrt) [inline tsensor/viz.py:111]: it looks like base color can change for same type depending on order in which this function is called, right?

Comment (parrt) [reply tsensor/viz.py:111]: I'll assign a specific color like "greenish" to floats or whatever.

Comment (parrt) [reply tsensor/viz.py:111]: BTW, i'm altering in a copy of your branch so i'll take a whack at fixing. I've made some changes like "no old style shape colors".

Comment (parrt) [reply tsensor/viz.py:111]: Ok I'm thinking we specify a fixed color per type then bit size is gradation for 4, 8, 16, 32, 64, 128 (auto computed with color map).  I'm looking at types for various packages.

**Comment by parrt** [comment]
I have to clean up AST viz but about ready!  Thanks for the help!

**Pull request merged**